### PR TITLE
[GH-993]: added support for subtasks in subscriptions

### LINKED ISSUE #859: Subtasks notification not working
I'm not getting the subtask notification in mattermost although I get all other notifications i.e Task, Story, Bug Epic.
it looks like there isn't any option of subtask in mattermost's Jira subscription

image

Mattermost Version: 7.0.1
Plugin version: 3.2

### LINKED ISSUE #993: Subtasks are not supported in subscription modal
At the moment, we don't support selecting "subtask types" as "issue types" in the subscription modal. It seems this is on purpose based on this line:

https://github.com/mattermost/mattermost-plugin-jira/blob/a86928d1d0c0c2493df5a2690ea8f54f515fd9a5/webapp/src/utils/jira_issue_metadata.tsx#L63

The task here is to look into the feasibility of supporting subtasks in the subscriptions feature. This would also affect the create issue modal if we were to change the line shown above.

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
#### Summary
Added functionality to support `subtasks` Jira subscription.

#### Old
Screenshot from 2024-07-04 11-26-52

#### New
Screenshot from 2024-06-28 18-04-05
Screenshot from 2024-06-28 17-16-43

#### Issues
  Fixes #993 
  Fixes #859

### CODE REVIEW

**Comment by Kshitij-Katiyar** [comment]
@mickmister Added the first part of the changes as per our discussion on https://github.com/mattermost/mattermost-plugin-jira/issues/993. Please have a look

File: webapp/src/utils/jira_issue_metadata.tsx [hunk 1]
```diff
@@ -55,7 +55,7 @@ export function getProjectValues(metadata: ProjectMetadata | null): ReactSelectO
     return metadata.projects;
 }
 
-export function getIssueTypes(metadata: IssueMetadata | null, projectKey: string | null): IssueType[] {
+export function getIssueTypes(metadata: IssueMetadata | null, projectKey: string | null, includeSubtasks: boolean): IssueType[] {
```
Comment (mickmister) [inline webapp/src/utils/jira_issue_metadata.tsx:58]: Just to make this easier to read on the caller's side

```suggestion
type GetIssueTypesOptions = {
    includeSubtasks: boolean;
}

export function getIssueTypes(metadata: IssueMetadata | null, projectKey: string | null, options: GetIssueTypesOptions): IssueType[] {
```

File: webapp/src/utils/jira_issue_metadata.tsx [hunk 2]
```diff
@@ -95,7 +100,7 @@ export function getFields(metadata: IssueMetadata | null, projectKey: string | n
         return {};
     }
 
-    const issueType = getIssueTypes(metadata, projectKey).find((it) => it.id === issueTypeId);
+    const issueType = getIssueTypes(metadata, projectKey, true).find((it) => it.id === issueTypeId);
```
Comment (mickmister) [inline webapp/src/utils/jira_issue_metadata.tsx:103]: Then this becomes

```suggestion
    const issueType = getIssueTypes(metadata, projectKey, {includeSubtasks: true}).find((it) => it.id === issueTypeId);
```

I trust your judgment on if this improves things

File: webapp/src/components/modals/create_issue/__snapshots__/create_issue_form.test.tsx.snap [hunk 3]
```diff
@@ -1377,7 +1377,7 @@ exports[`components/CreateIssue should match snapshot with error 1`] = `
                     "self": "https://mmtest.atlassian.net/rest/api/2/issuetype/10002",
                   },
                   Object {
-                    "description": "The sub-task of the issue",
+                    "description": "Stories track functionality or features expressed as user goals.",
```
Comment (mickmister) [inline webapp/src/components/modals/create_issue/__snapshots__/create_issue_form.test.tsx.snap:1380]: Do you know why this snapshot changed so much?

**Review by mickmister — APPROVED** [review]

File: webapp/src/components/modals/create_issue/__snapshots__/create_issue_form.test.tsx.snap (hunk 3 above)
Comment (Kshitij-Katiyar) [reply webapp/src/components/modals/create_issue/__snapshots__/create_issue_form.test.tsx.snap:1380]: @mickmister some changes are related to the subtask and apart from that the order of file content is shuffled, mostly all the existing test cases are present in the new changes.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:

**Review by AayushChaudhary0001 — COMMENTED** [review]
@raghavaggarwal2308  found a issue while testing this PR.
Issue: After creating the subtask, added subscriptions are not shown.

Step to reproduce:

1. Create a new JIRA subscription for any channel
2. Select the subtask for that subscription
3. Create a new subtask
4. Check the bot message on the channel

Expected: Subscription part is present in the bot message
Actual: Subscription part is missing in the bot message

Note: Mattermost Version: 8.1.0

**Comment by raghavaggarwal2308** [comment]
@AayushChaudhary0001 You have to enable the below setting in system console to see that:
image

**Review by AayushChaudhary0001 — APPROVED** [review]
Working fine, Approved

**Pull request merged**